src/NSubstitute/SubstituteExtensions.cs

```diff
@@ -354,5 +371,21 @@ private static ValueTask<T> CompletedValueTask<T>(T result)
         {
             return new ValueTask<T>(result);
         }
+
+        private static void ReThrowIfFaulted<T>(Task<T> task)
+        {
+            if (task.IsFaulted && task.Exception.InnerExceptions[0] is SubstituteException)
```

Comment:
`InnerExceptions` will never be empty?

Reply:
The `Exception` property is of `AggregateException` type. I presume, it's impossible to create `AggregateException` which aggregates nothing. I think realistically it should never be null for faulted task.

If you believe we should play extra-safe here, I could, of course, replace it with `FirstOrDefault()`...

Reply:
I think safe is better, may need to handle `null` too (see [`AggregateException(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.aggregateexception.-ctor?view=netframework-4.8#System_AggregateException__ctor_System_String_)).

Reply:
@dtchepak Done!

Reply:
... While I still believe it's impossible to craft a task with `InnerExceptions` set to null 🤔 If you manage to do it - could you share a sample?

Anyway, let's keep `FirstOrDefault` 👍

Reply:
Does this count?
```csharp
var ex = (AggregateException)FormatterServices.GetUninitializedObject(typeof(AggregateException));
```
😝 

Looks like from [source](https://referencesource.microsoft.com/#mscorlib/system/AggregateException.cs) this should never be null in normal usage (with current implementation at least).